docbrown/db/src/algorithms/reciprocity.rs

```diff
@@ -0,0 +1,97 @@
+use itertools::Itertools;
+use crate::graph_window::{WindowedGraph, WindowedVertex};
+
+
+fn get_unique_out_neighbours(v: &WindowedVertex) -> Vec<u64> {
+    let mut all_out_neighbours: Vec<u64> = v.out_neighbours_ids().collect();
+    // remove self-loop
+    if let Some(pos) = all_out_neighbours.iter().position(|x| *x == v.g_id) {
+        all_out_neighbours.remove(pos);
+    }
+    all_out_neighbours
+}
+
+fn get_unique_in_neighbours(v: &WindowedVertex) -> Vec<u64> {
+    let mut in_neighbours: Vec<u64>  = v.in_neighbours_ids().collect();
+    // remove self-loop
+    if let Some(pos) = in_neighbours.iter().position(|x| *x == v.g_id) {
+        in_neighbours.remove(pos);
+    }
+    in_neighbours
+}
+
+fn get_reciprocal_edge_count(v: &WindowedVertex) -> usize {
+    let out_neighbours: Vec<u64> = get_unique_out_neighbours(&v);
+    let in_neighbours: Vec<u64> = get_unique_in_neighbours(&v);
+    out_neighbours.iter().filter(|x| in_neighbours.contains(x)).count()
```

Comment:
think this operation is O(n^2) as the vec `contains` method is O(n), would it be better to do this as a hash set intersect?

Reply:
good shout, done

Reply:
good shout, done